src/asl.ts

```diff
@@ -5526,28 +5253,1768 @@ export namespace ASLGraph {
+              type: "PropAssignment",
+              from: parameters as string,
+              to: resultPath,
+              props,
+            },
+      ];
+    }
+
+    /**
+     * Pull variable references out of intrinsic functions and json path filters.
+     *
+     * States.Array($.var) => $.var
+     * $.arr[?(['value'] == $.var)] => $.arr, $.var
+     */
+    function extractVariableReferences(jsonPath: string) {
+      return jsonPath.match(/(\$\.?[^,)]+)+/g) ?? [];
+    }
+  }
+
+  function getVariableNamesFromState(state: State): string[] {
+    const names = ((state) => {
+      if (
+        isTaskState(state) ||
+        isMapTaskState(state) ||
+        isParallelTaskState(state)
+      ) {
+        return [
+          state.ResultPath,
+          ...(state.Catch?.map((c) => c.ResultPath) ?? []),
+        ];
+      } else if (isPassState(state)) {
+        return [state.ResultPath];
+      } else if (
+        isSucceedState(state) ||
+        isFailState(state) ||
+        isChoiceState(state) ||
+        isWaitState(state)
+      ) {
+        return [];
+      }
+      return assertNever(state);
+    })(state);
+
+    return names.filter((n): n is string => !!n);
+  }
+
+  export const DefaultOptimizeOptions: OptimizeOptions = {
+    optimizeVariableAssignments: true,
+    removeUnreachableStates: true,
+    joinConsecutiveChoices: true,
+    removeNoOpStates: true,
+  };
+
+  export interface OptimizeOptions {
+    /**
+     * Attempts to collapse assignments that provide no functional value.
+     *
+     * Collapses assignments created by the interpreter and the input source.
+     *
+     * ```ts
+     * const a = "a"
+     * const b = a;
+     * =>
+     * const b = "a"
+     * ```
+     */
+    optimizeVariableAssignments: boolean;
+    /**
+     * Traverses the graph transitions and removes any states that cannot be reached.
+     * Unreachable states will fail on deployment to AWS Step Functions.
+     *
+     * Note: If this optimization is turned off, it is likely the output ASL will fail.
+     *       Functionless creates some unreachable that this logic is expected to remove.
+     *       Only turn this off if you intend to preform similar logic or use the ASL for another
+     *       purpose than AWS Step Functions.
+     */
+    removeUnreachableStates: boolean;
+    /**
+     * When {@link Choice} states are chained together, they can be simplified by merging the branches together.
+     *
+     * ```ts
+     * if(a) {
+     *   if(b) {
+     *   } else {
+     *   }
+     * }
+     *
+     * =>
+     *
+     * ```ts
+     * if(a && b) {}
+     * else(a) {}
+     * ```
+     *
+     * In ASL this may remove unnecessary state transitions without any impact on the logic.
+     *
+     * Note: This could make the ASL itself larger when the same Choice is merged into multiple calling choices.
+     *       Turn off at the cost of adding more state transitions.
+     */
+    joinConsecutiveChoices: boolean;
```

Comment:
Love it!

Do we have a test case where there are subsequent choices except the inner one has an else?

```ts
if(a) {
  if(b) {}
  else {}
}
```

Reply:
(this was in there before)

We did not, added one

```ts
      if (input.val !== "a") {
        if (input.val === "b") {
          return "hullo";
        }
      }
      return "woop";
```

```ts
    "if(input.val !== \"a\")": Object {
      "Choices": Array [
        Object {
          "And": Array [
            Object {
              "Not": Object {
                "And": Array [
                  Object {
                    "IsPresent": true,
                    "Variable": "$.input.val",
                  },
                  Object {
                    "And": Array [
                      Object {
                        "IsString": true,
                        "Variable": "$.input.val",
                      },
                      Object {
                        "StringEquals": "a",
                        "Variable": "$.input.val",
                      },
                    ],
                  },
                ],
              },
            },
            Object {
              "And": Array [
                Object {
                  "IsPresent": true,
                  "Variable": "$.input.val",
                },
                Object {
                  "And": Array [
                    Object {
                      "IsString": true,
                      "Variable": "$.input.val",
                    },
                    Object {
                      "StringEquals": "b",
                      "Variable": "$.input.val",
                    },
                  ],
                },
              ],
            },
          ],
          "Next": "return \"hullo\"",
        },
        Object {
          "Next": "return \"wat\"",
          "Not": Object {
            "And": Array [
              Object {
                "IsPresent": true,
                "Variable": "$.input.val",
              },
              Object {
                "And": Array [
                  Object {
                    "IsString": true,
                    "Variable": "$.input.val",
                  },
                  Object {
                    "StringEquals": "a",
                    "Variable": "$.input.val",
                  },
                ],
              },
            ],
          },
        },
      ],
      "Default": "return \"woop\"",
      "Type": "Choice",
    },
```